artemis-server/src/main/resources/schema/artemis-configuration.xsd

```diff
@@ -675,6 +675,82 @@
             </xsd:annotation>
          </xsd:element>
 
+         <xsd:element name="journal-retention" maxOccurs="1" minOccurs="0">
+            <xsd:annotation>
+               <xsd:documentation>
+                  Configuration about the journal retention configuration.
+               </xsd:documentation>
+            </xsd:annotation>
+            <xsd:complexType>
+               <xsd:attribute name="unit" use="required">
+                  <xsd:annotation>
+                     <xsd:documentation>
+                        This configures the period type to use on limit. By default it is DAYS.
+                     </xsd:documentation>
+                  </xsd:annotation>
+                  <xsd:simpleType>
+                     <xsd:restriction base="xsd:string">
+                        <xsd:enumeration value="DAYS"/>
+                        <xsd:enumeration value="HOURS"/>
+                        <xsd:enumeration value="SECONDS"/>
+                     </xsd:restriction>
+                  </xsd:simpleType>
+               </xsd:attribute>
+               <xsd:attribute name="directory" type="xsd:string" use="required">
```

Comment:
Why is this an attribute but for all other data directories its the element value? Just a consistency thing.

Reply:
the element is journal-retention.. you simply enable it or disable it...

I think it is very neat the way it's done now.


you just comment this from the xml, and you have it working:

```xml
<journal-retention unit="DAYS" directory="data/retention" period="7" storage-limit="10G"/>
```

Reply:
Yes but other dirs the element value is the directory name... not saying to move attributes for the retention parts. Just saying shouldnt the directory attribute be the element value as like other data dirs

Reply:
Eg

`<journal-retention unit="DAYS" period="7" storage-limit="10G">data/retention</journal-retention>`

Reply:
you mean this?

```
<journal-retention unit="DAYS" period="7" storage-limit="10G">data/retention</journal-retention>
```

I would have to call it journal-retention-directory.. it wouldn't look as nice IMO. I would probably have to split it into separate elements... I would rather keep the way it is now.

Reply:
Im one for consistency. Especially as a user mindset. Like why all other dirs the element value yet here its an empty element value but its an attribute.

Reply:
It is not possible to have both attributes and default children on a xml type. I will go with what I had before... and I need to move on on this.

Reply:
I dont follow. Im sure if its enabled by declaring the xml were ensuring that directory is NOT optional.

Reply:
the other data dirs don't have attributes. I don't want to get stuck over this TBH.

Reply:
I can't make it this on the xsd:


```xml
<journal-retention-directory period="x">directory</journal-retention-directory
```


There are other complex types on the XSD for the configuration.. I don't see an issue with this TBH... And I really need to finish this task as I need to finish the release.

Reply:
Its about consistency. And the issue is once released we cant easily change it.

We actually already have such type on xsd/xml. connectors and acceptors which have a value as well as attributes.

Reply:
I believe sample type in the xsd is transportType

Reply:
```
            <xsd:element name="journal-retention-directory" maxOccurs="1" minOccurs="0">
                <xsd:annotation>
                    <xsd:documentation>
                        the directory to store journal-retention message in and rention configuraion.
                    </xsd:documentation>
                </xsd:annotation>
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="unit" use="optional">
                                <xsd:annotation>
                                    <xsd:documentation>
                                        This configures the period type to use on limit. By default it is DAYS.
                                    </xsd:documentation>
                                </xsd:annotation>
                                <xsd:simpleType>
                                    <xsd:restriction base="xsd:string">
                                        <xsd:enumeration value="DAYS"/>
                                        <xsd:enumeration value="HOURS"/>
                                        <xsd:enumeration value="SECONDS"/>
                                    </xsd:restriction>
                                </xsd:simpleType>
                            </xsd:attribute>
                            <xsd:attribute name="period" type="xsd:integer" use="optional">
                                <xsd:annotation>
                                    <xsd:documentation>
                                        The amount of time used to keep files.
                                    </xsd:documentation>
                                </xsd:annotation>
                            </xsd:attribute>
                            <xsd:attribute name="storage-limit" type="xsd:string" use="optional">
                                <xsd:annotation>
                                    <xsd:documentation>
                                        Size (in bytes) before we starting removing files from the retention area.
                                        this is an extra protection on top of the period.
                                        Notice we first remove files based on period and if you're using more storage then you
                                        configured we start removing older files.
                                        By default this is unlimited (not filled).
                                        Supports byte notation like "K", "Mb", "GB", etc.
                                    </xsd:documentation>
                                </xsd:annotation>
                            </xsd:attribute>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>```

Reply:
dont forget to add    <xsd:attributeGroup ref="xml:specialAttrs"/>

Reply:
<xsd:element name="journal-retention-directory" maxOccurs="1" minOccurs="0">
                <xsd:annotation>
                    <xsd:documentation>
                        the directory to store e journal-retention will take place.
                    </xsd:documentation>
                </xsd:annotation>
                <xsd:complexType>
                    <xsd:simpleContent>
                        <xsd:extension base="xsd:string">
                            <xsd:attribute name="unit" use="optional" type="timeUnit">
                                <xsd:annotation>
                                    <xsd:documentation>
                                        This configures the period type to use on limit. By default it is DAYS.
                                    </xsd:documentation>
                                </xsd:annotation>
                            </xsd:attribute>
                            <xsd:attribute name="period" type="xsd:integer" use="optional">
                                <xsd:annotation>
                                    <xsd:documentation>
                                        The amount of time used to keep files.
                                    </xsd:documentation>
                                </xsd:annotation>
                            </xsd:attribute>
                            <xsd:attribute name="storage-limit" type="xsd:string" use="optional">
                                <xsd:annotation>
                                    <xsd:documentation>
                                        Size (in bytes) before we starting removing files from the retention area.
                                        this is an extra protection on top of the period.
                                        Notice we first remove files based on period and if you're using more storage then you
                                        configured we start removing older files.
                                        By default this is unlimited (not filled).
                                        Supports byte notation like "K", "Mb", "GB", etc.
                                    </xsd:documentation>
                                </xsd:annotation>
                            </xsd:attribute>
                        </xsd:extension>
                    </xsd:simpleContent>
                </xsd:complexType>
            </xsd:element>


And also define timeUnit so its re-usable going forwards, after all its java TimeUnit that we use to convert.

    <xsd:simpleType name="timeUnit">
        <xsd:restriction base="xsd:string">
            <xsd:enumeration value="DAYS"/>
            <xsd:enumeration value="HOURS"/>
            <xsd:enumeration value="MINUTES"/>
            <xsd:enumeration value="SECONDS"/>
        </xsd:restriction>
    </xsd:simpleType>

Reply:
ok.. thanks.. that works